### PR TITLE
User management system

### PR DESCRIPTION
**Pull request by AVaitkunas**
User management system.

### CODE REVIEW

File: pkg/database/models.go [hunk 1]
```diff
@@ -4,9 +4,18 @@ import (
...
+	"github.com/gookit/validate"
+	"golang.org/x/crypto/bcrypt"
 	"gorm.io/gorm"
 )
 
+type RoleType string
+
+const (
+	RoleUser  RoleType = "user"
+	RoleAdmin RoleType = "admin"
```
Comment (AngelVI13) [inline pkg/database/models.go]: The if the type is the same, the second line doesnt need to have it
```suggestion
	RoleAdmin  = "admin"
```

File: pkg/database/models.go [hunk 2]
```diff
@@ -26,17 +35,61 @@ type Keyword struct {
...
+	Password string `form:"password" validate:"required"`
+}
+
+func (l LoginForm) Messages() map[string]string {
+	return validate.MS{
+		"required": "Field {field} is required!",
+	}
 }
+
+func (l LoginForm) CheckLogin(db *gorm.DB) (User, error){
```
Comment (AngelVI13) [inline pkg/database/models.go]: Consider returning pointer to User
```
func (l LoginForm) CheckLogin(db *gorm.DB) (User, error){
```

File: pkg/routes/context.go [hunk 3]
```diff
@@ -31,6 +32,18 @@ func (c *Ctx) WithUrls() *Ctx {
 	return c
 }
 
+func (c *Ctx) WithSession() *Ctx {
+	data := fiber.Map{}
+
+	user, _ := session.GetActiveUsername(c.Ctx)
```
Comment (AngelVI13) [inline pkg/routes/context.go]: Understand why this handling is not needed and maybe add a comment explaining why its not handled
```
	user, _ := session.GetActiveUsername(c.Ctx)
```

File: pkg/routes/login.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+)
+
+func (r *Router) HandleLoginGet(c *Ctx) error {
+	data := c.FlashData()
+	data["Title"] = "Login"
+	return c.Render(LoginView, data)
+}
+
+func (r *Router) HandleLoginPost(c *Ctx) error {
+	data := c.FlashData()
```
Comment (AngelVI13) [inline pkg/routes/login.go]: No need for flash data in Post method
```
	data := c.FlashData()
```

File: pkg/routes/login.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+	data["Title"] = "Login"
+	
+	var login database.LoginForm
+	err := c.BodyParser(&login)
+
+	if err != nil {
+		return c.WithError(
+			fmt.Sprintf("credentials validation failed. Error: %s", err.Error()),
```
Comment (AngelVI13) [inline pkg/routes/login.go]: This error is related to body parsing and not to credentials validation
```
			fmt.Sprintf("credentials validation failed. Error: %s", err.Error()),
```

File: pkg/routes/login.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+import (
+	"fmt"
+
+	"github.com/AngelVI13/fiber-investigation/pkg/database"
+	"github.com/AngelVI13/fiber-investigation/pkg/session"
+	"github.com/gookit/validate"
+)
+
+func (r *Router) HandleLoginGet(c *Ctx) error {
```
Comment (AngelVI13) [inline pkg/routes/login.go:11]: Do not show login page if user is already logged in
```
func (r *Router) HandleLoginGet(c *Ctx) error {
```

File: pkg/routes/login.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+	// validate input data
+	v := validate.Struct(user)
+	if !v.Validate() {
+		return c.WithError(
+			fmt.Sprintf("failed to register new user. Error: %s", v.Errors),
+		).RedirectBack(IndexUrl)
+	}
+
+	user.HashPassword()
```
Comment (AngelVI13) [inline pkg/routes/login.go]:
```suggestion
	hashedPass := HashPassword(user.Password)
	user.SetPassword(hashedPass)
```

File: pkg/routes/login.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+	result := r.db.Create(&user)
+
+	if result.Error != nil {
+		return c.WithError(
+			fmt.Sprintf("failed to register new user. Error: %s", result.Error),
+		).RedirectBack(IndexUrl)
+	}
+
+	defer session.Login(c.Ctx, user)
```
Comment (AngelVI13) [inline pkg/routes/login.go]: This is hiding any error that might be raised by seesion.Login
```
	defer session.Login(c.Ctx, user)
```

File: pkg/session/session_operations.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+)
+
+func Login(c *fiber.Ctx, user database.User) error {
+	session, err := SessionStore.Get(c)
+
+	if err != nil {
+		return fmt.Errorf("failed to get session. Error %s", err.Error())
+	}
+
+	username, _ := GetActiveUsername(c)
```
Comment (AngelVI13) [inline pkg/session/session_operations.go]: Username here can't be null:
1. Update GetActiveUsername to return a pointer to user
2. Check the error for failure case and not the username
```
	username, _ := GetActiveUsername(c)
```

File: pkg/session/session_operations.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+	return nil
+}
+
+func GetActiveUsername(c *fiber.Ctx) (interface{}, error) {
+	session, err := SessionStore.Get(c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get session. Error %s", err.Error())
+	}
+
+	return session.Get(SessionUsername), nil
```
Comment (AngelVI13) [inline pkg/session/session_operations.go]: Do not return an interface but instead cast it to string and handle error
```
	return session.Get(SessionUsername), nil
```

File: pkg/session/session_operations.go [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+func IsUserActive(c *fiber.Ctx) bool {
```
Comment (AngelVI13) [inline pkg/session/session_operations.go]: Update to handle case of error and user != ""
```
func IsUserActive(c *fiber.Ctx) bool {
```

File: views/keywords.html [hunk 12]
```diff
@@ -10,6 +10,8 @@ <h2>{{.Title}}</h2>
             <th>Arguments</th>
             <th>Documentation</th>
             <th>Implementation</th>
+            {{if and $.IsAdmin $.IsConnected}}
+            {{if eq $.LatestVersion $.SelectedVersion}}
```
Comment (AngelVI13) [inline views/keywords.html]: Indent this a bit more for clarity
```
            {{if eq $.LatestVersion $.SelectedVersion}}
```

File: pkg/database/models.go (hunk 1 above)
Comment (AVaitkunas) [reply pkg/database/models.go]: fixed

File: pkg/database/models.go (hunk 2 above)
Comment (AVaitkunas) [reply pkg/database/models.go]: fixed

File: pkg/routes/context.go (hunk 3 above)
Comment (AVaitkunas) [reply pkg/routes/context.go]: fixed

File: pkg/routes/login.go (hunk 6 above)
Comment (AVaitkunas) [reply pkg/routes/login.go:11]: fixed

File: pkg/routes/login.go (hunk 4 above)
Comment (AVaitkunas) [reply pkg/routes/login.go]: fixed

File: pkg/routes/login.go (hunk 5 above)
Comment (AVaitkunas) [reply pkg/routes/login.go]: fixed

File: pkg/routes/login.go (hunk 7 above)
Comment (AVaitkunas) [reply pkg/routes/login.go]: fixed

File: pkg/routes/login.go (hunk 8 above)
Comment (AVaitkunas) [reply pkg/routes/login.go]: fixed

File: pkg/session/session_operations.go (hunk 9 above)
Comment (AVaitkunas) [reply pkg/session/session_operations.go]: fixed

File: pkg/session/session_operations.go (hunk 10 above)
Comment (AVaitkunas) [reply pkg/session/session_operations.go]: fixed

File: pkg/session/session_operations.go (hunk 11 above)
Comment (AVaitkunas) [reply pkg/session/session_operations.go]: fixed

File: views/keywords.html (hunk 12 above)
Comment (AVaitkunas) [reply views/keywords.html]: fixed

**Review by AngelVI13 — APPROVED** [review]

**Pull request merged**